### PR TITLE
proposal: Live migration for bridged pod network

### PR DESCRIPTION
**Pull request by kvaps**
- Rendered version: [live-migration-for-bridged-pod-network.md](https://github.com/kvaps/community/blob/live-migration-for-bridged-pod-network/design-proposals/live-migration-for-bridged-pod-network.md)
- Implementation: https://github.com/kubevirt/kubevirt/pull/7768
- Slack thread:
https://kubernetes.slack.com/archives/C8ED7RKFE/p1652296916563589?thread_ts=1650371688.593359&cid=C8ED7RKFE

### CODE REVIEW

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (maiqueb) [inline design-proposals/live-migration-for-bridged-pod-network.md:44]:
```suggestion
- Wait for launch
```

Now I'm hungry :smiley:

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (maiqueb) [inline design-proposals/live-migration-for-bridged-pod-network.md:29]: As you indicate in your non-goal section this will impact workloads listening on particular IP addresses. 

I think this might be OK if we protect this feature behind some kind of feature gate (disabled by default), and isolate well the flow.

Your design should - imo - address that.

**Review by maiqueb — CHANGES_REQUESTED** [review]
Thanks for taking the time to come up w/ this proposal.

I'm mostly worried / concerned about the workloads that are bound to certain addresses.

I'm unsure we can accept this in without offering a way to protect those.

@AlonaKaplan would you review this proposal as well ?

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (rthallisey) [inline design-proposals/live-migration-for-bridged-pod-network.md:12]: Are these assumptions and limitations?  They don't seem like Non-goals. 

Macvtap has the limitation on the [website](https://kubevirt.io/user-guide/virtual_machines/interfaces_and_networks/#macvtap): "Live migration is not seamless, see [issue #5912](https://github.com/kubevirt/kubevirt/issues/5912#issuecomment-888938920)".    You probably want something similar.

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (rthallisey) [inline design-proposals/live-migration-for-bridged-pod-network.md:39]: Can you reword this?  Do you mean to say that when using multiple NICs, there are some CNIs that didn't work with live-migration.  Now these CNIs will work.

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (kvaps) [reply design-proposals/live-migration-for-bridged-pod-network.md:2]: mentions of macvtap has been removed

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (kvaps) [reply design-proposals/live-migration-for-bridged-pod-network.md:29]: added to design

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (kvaps) [reply design-proposals/live-migration-for-bridged-pod-network.md:12]: Thanks, added to the Non goals

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (kvaps) [reply design-proposals/live-migration-for-bridged-pod-network.md:39]: added thanks!

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (EdDev) [inline design-proposals/live-migration-for-bridged-pod-network.md:25]: - I think the name better be specific here, e.g. `DisruptiveBridgeBindingMigration`.
- Could you please elaborate on when in the (migration) flow you plan to execute the detach/attach operation?
- Could you please elaborate on when in the (migration) flow you plan to execute the link down/up operation?
- If I understand correctly, the IP address is always expected to change, no matter if the mac is changed or not. Right?

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (EdDev) [inline design-proposals/live-migration-for-bridged-pod-network.md:6]: I would be interested to understand the motivation to have a disruptive migration.
If such a migration can cause up to a minute of disruption, aren't there other means to recover a VM on a different node? I was under the impression that live-migration with high disruption of connectivity can cause more harm then good (applications will still run while disruption occurs, vs applications with the whole VM freeze and get restored).

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (EdDev) [inline design-proposals/live-migration-for-bridged-pod-network.md:52]: There is the topic of maintaining this special flow.
Unfortunately, our code is not well fitted to introduce a new option and at the same time keep it isolated and centralized. When the additions are scattered across many areas, it makes it harder to maintain it.

Well, it may be well worth investing in maintaining it by the sig-network if it has enough attraction, interest and at the end real usage.
I think it is worth raising these points now, so we could be clear that even if this feature is accepted as an Alpha, its existence for the long run is dependent on adoption and usage. I guess this is similar to the process of Kuberenetes, but we should have one as well, especially for such controversy features.

**Review by EdDev — COMMENTED** [review]
Thank you for this interesting proposal!

I like it and at the same time a bit worried.
It is not very clear to me how much potential adoption it may have due to the disruptions and limitations. But perhaps this is a good opportunity to try it and see.

I have commented inline about a few questions and concerns.

Thank you for bringing this to the table.

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (kvaps) [reply design-proposals/live-migration-for-bridged-pod-network.md:25]: > I think the name better be specific here, e.g. DisruptiveBridgeBindingMigration.

I wouldn't specify any specific mode in feature gate name. The [next PR](https://github.com/kubevirt/kubevirt/pull/7648) is upcoming, it will bring macvtap binding mode which also will work with this feature.

> - Could you please elaborate on when in the (migration) flow you plan to execute the detach/attach operation?
> - Could you please elaborate on when in the (migration) flow you plan to execute the link down/up operation?

That was described here

https://github.com/kubevirt/community/blob/eb0d5dddf8aac2ba5e817e5a79b5ef9e7aa54812/design-proposals/live-migration-for-bridged-pod-network.md?plain=1#L26-L32

If you talking about when exactly is it going to happen, sure, I will add the information that this will happened after the successful live-migration


> If I understand correctly, the IP address is always expected to change, no matter if the mac is changed or not. Right?
Not exactly, some CNIs allow to specify IP-address in annotation, this way IP-address will not be changed, but we still need to link down and link up the interface to make sure that VM has correct routes.

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (kvaps) [reply design-proposals/live-migration-for-bridged-pod-network.md:6]: > If such a migration can cause up to a minute of disruption, aren't there other means to recover a VM on a different node?

From my tests this is happening within a few seconds not minutes, so even tcp sessions do not have time to break.

Comment (davidvossel) [reply design-proposals/live-migration-for-bridged-pod-network.md:6]: how do tcp connections survive an ip address change?

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (maiqueb) [reply design-proposals/live-migration-for-bridged-pod-network.md:18]: I would actually be more interested in what apps / scenarios do **not** tolerate this kind of live migration.

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (maiqueb) [reply design-proposals/live-migration-for-bridged-pod-network.md:50]: likewise. I share your expectation, and I think all the established tcp connections will go down.

**Comment by AlonaKaplan** [comment]
Hi @kvaps thank you very much for the proposal and the PR. The idea generally looks good.
We're currently working on moving the binding mechanism outside of kubevirt and implementing each binding as a separate CNI.
This will allow to easily integrate/update new/existing bindings.
Until this change is introduced, we try to avoid introducing dramatic changes to the existing bindings.

@maiqueb is currently working on it and we will update this PR once moving the bridge binding to a CNI will be ready.

**Comment by maiqueb** [comment]
> Hi @kvaps thank you very much for the proposal and the PR. The idea generally looks good. We're currently working on moving the binding mechanism outside of kubevirt and implementing each binding as a separate CNI. This will allow to easily integrate/update new/existing bindings. Until this change is introduced, we try to avoid introducing dramatic changes to the existing bindings.
> 
> @maiqueb is currently working on it and we will update this PR once moving the bridge binding to a CNI will be ready.

I want to make it clear that you should adjust your expectations: I started to look into this **very** recently, and currently can't provide a time-line for it. I'll CC you both on the design document once I have a plan set up for it. Unfortunately, this is **not** my current primary focus, meaning it will take some time.

The motivation behind it is to enable **you** (and anyone interested in providing another binding type / custom binding behavior) to do so, experiment with it, while keeping the maintainability of the KubeVirt project to something manageable.

I would welcome any sort of input / collaboration for this task, going forward @kvaps .

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (kvaps) [reply design-proposals/live-migration-for-bridged-pod-network.md:6]: Some CNIs support specifying IP-address per pod. So in this way IP-address is not changed, but still requires to renew DHCP lease for updating routes inside the VM

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (kvaps) [reply design-proposals/live-migration-for-bridged-pod-network.md:50]: There is same approach we having in test for macvtap interface. It's not the same ping command, but two different:

eg:

- Run ping from client to server (client has an IP 10.10.10.1, server has IP 10.10.10.20, server responding to 10.10.10.1)
- live-migrate client VM (client has an IP 10.10.10.2, server has IP 10.10.10.20, ping is not running)
- Run ping from client to server (client has an IP 10.10.10.2, server has IP 10.10.10.20, server responding to 10.10.10.2)

**Comment by kvaps** [comment]
@AlonaKaplan @maiqueb this PR does not bring are new binding modes to KubeVirt, its just allowing to live-migrate in case if you use bridge.
Maybe you talk about another my PR kubevirt/kubevirt#7648 or do you mean both of them?

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (AlonaKaplan) [inline design-proposals/live-migration-for-bridged-pod-network.md:5]: To be honest, I never thought about performance as a reason to use bridge binding on the primary network.
If you need better performance, why not using a secondary network?
Can you please elaborate about the use case.

Comment (kvaps) [reply design-proposals/live-migration-for-bridged-pod-network.md:5]: We have no other options to expose VM as Kubernetes service.
We still need to communicate with the other pods in the cluster and reuse cloud load-balancers or MetalLB.
We forced to use pod network in any case we want to communicate with Kubernetes.

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (kvaps) [reply design-proposals/live-migration-for-bridged-pod-network.md:18]: > This may affect some applications which are binding to the specific interface inside the VM.

Isn't it enough? Or should I specify cases with exact applications (eg. apache2 server configured to bind on pod IP instead of `0.0.0.0` and so on)?

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (AlonaKaplan) [reply design-proposals/live-migration-for-bridged-pod-network.md:6]: Did you try to use masquerade with multiqueue?
That's the result of a comparison we did in the past (masquerade binding, 8vcpu 8multiqueue)-

```
Masquerade overhead is minimized with multiqueue.
Masquerade vs. bridge binding
TCP Stream: -37% (nomq) -13% (mq)
UDP Stream:  -2% (nomq) -6% (mq)
TCP Latency: -24% (nomq) -3% (mq)
UDP Latency: -6% (nomq) -2% (mq)
```

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (AlonaKaplan) [reply design-proposals/live-migration-for-bridged-pod-network.md:5]: Did you try to use masquerade with multiqueue?
That's the result of a comparison we did in the past (masquerade binding, 8vcpu 8multiqueue)-

Masquerade overhead is minimized with multiqueue.
Masquerade vs. bridge binding
TCP Stream: -37% (nomq) -13% (mq)
UDP Stream:  -2% (nomq) -6% (mq)
TCP Latency: -24% (nomq) -3% (mq)
UDP Latency: -6% (nomq) -2% (mq)

**Comment by AlonaKaplan** [comment]
> @AlonaKaplan @maiqueb this PR does not bring are new binding modes to KubeVirt, its just allowing to live-migrate in case if you use bridge. Maybe you talk about another my PR [kubevirt/kubevirt#7648](https://github.com/kubevirt/kubevirt/pull/7648) or do you mean both of them?

This and the other PR made us think about it. And we realized there is a need to allow easily adding new bindings/extend the existing ones.

Bridge binding on the primary network is a legacy binding. We still support it, we are fixing bugs on it. But we try to avoid adding new functionality to it.
For example, IPv6 support was never added to it.
We encourage users to use masquerade as the primary binding. If better performance is needed masquerade can be used with multi queues.

However, once each binding will be implemented as a separate CNI it will be possible to add whatever binding CNI you need or extend an existing one - without the need to get an approval from the kubevirt team.

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (EdDev) [reply design-proposals/live-migration-for-bridged-pod-network.md:50]: - The server is not the one migrated, it is the client (line 49). So initiating a ping from the client should work fine.
- The test is on ICMP packets, not TCP.
- This is not much different from masquerade binding tests. In masquerade binding, it visible IP is also changing, so this is not something new.
- TCP connections are dropped/reset in masquerade binding as well, our existing migration is not smooth as well.

I would add here TCP/UDP tests as well, including ones with a service (to show that even if the IP changes, access through the service is returned eventually).

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (EdDev) [reply design-proposals/live-migration-for-bridged-pod-network.md:6]: They won't survive. It will not survive even a link flickering... those TCP connections will just reset.
But this is the case with masquerade as well, so this is not something new.
I was more worried about the duration of the downtime (e.g. in the case of attach/detach).

**Comment by AlonaKaplan** [comment]
@kvaps throwing another idea that you may use until the CNI binding mechanism will be introduced.
We have a [hook point](https://github.com/kubevirt/kubevirt/blob/25b37338917f1a7bfbe3fc07d87672ef1e39389c/pkg/virt-launcher/virtwrap/live-migration-target.go#L92) for additional setup on the target pod.
You can add the plug/unlug/link/unlink code to a hook sidecar.

**Comment by kvaps** [comment]
Thanks, I'll take a look at this, but since we want to have both functions I think it might be more reasonable to support our own fork with proposed functionality until it will be possible to implement as module.

**Comment by EdDev** [comment]
> Thanks, I'll take a look at this, but since we want to have both functions I think it might be more reasonable to support our own fork with proposed functionality until it will be possible to implement as module.

The modularity work is in the planning, so I wanted to check if it makes also sense to support pluggability at build time.
Unlike the CNI option, this other option is just a refactoring effort to generalize the flow and structuring it in a way arbitrary bindings and behaviors can get registered at build time.
Is this something that can be of interest? Even if you are forking, it will make rebases easier (as long as we do not break to often that pluggable API).

**Comment by rthallisey** [comment]
Any updates on this @kvaps?  Are you waiting for any more feedback?

**Comment by kvaps** [comment]
Hi @rthallisey, sorry for late reply.
I'm currently preparing my new devel stand for KubeVirt, when I finish I'll answer to all questions in this PR.
I want to make some tests for that

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (kvaps) [reply design-proposals/live-migration-for-bridged-pod-network.md:6]: Well I just performed a test, I used the flowing commands:

server vm:
```
mkdir cgi-bin
printf '%s\n' '#!/bin/sh' 'cat /dev/urandom' > cgi-bin/1.sh
chmod +x cgi-bin/1.sh
python -m http.server --cgi
```

client vm:
```
curl 10.10.10.1:8000/cgi-bin/1.sh > /dev/null
```

Now I can assure that in case if IP address is not changed the tcp-connections are surviving


Also I checked flood ping in both cases. When MAC-address is changes and when is not, result:

**When MAC-address is the same (link down and link up interface):**

<details>
<summary>
47 packets lost
</summary>
<pre>
[fedora@vm1 ~]$ sudo ping -f 10.10.10.2
PING 10.10.10.2 (10.10.10.2) 56(84) bytes of data.
...............................................
--- 10.10.10.2 ping statistics ---
72544 packets transmitted, 72497 received, 0.0647883% packet loss, time 14943ms
rtt min/avg/max/mdev = 0.069/0.117/8.138/0.160 ms, ipg/ewma 0.205/0.106 ms
</pre>
</details>

**When MAC-address is changed (reattach network card):**  

<details>
<summary>
54 packets lost
</summary>
<pre>
[fedora@vm1 ~]$ sudo ping -f 10.10.10.2
PING 10.10.10.2 (10.10.10.2) 56(84) bytes of data.
......................................................
--- 10.10.10.2 ping statistics ---
102443 packets transmitted, 102389 received, 0.0527122% packet loss, time 20785ms
rtt min/avg/max/mdev = 0.069/0.116/39.293/0.222 ms, pipe 4, ipg/ewma 0.202/0.355 ms
</pre>
</details>

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (kvaps) [reply design-proposals/live-migration-for-bridged-pod-network.md:5]: @AlonaKaplan which tool you used to benchmark this?

**Comment by kvaps** [comment]
@rthallisey I answered the most of conversation threads, not sure, what I suppose to do now?

File: design-proposals/live-migration-for-bridged-pod-network.md
Comment (rthallisey) [reply design-proposals/live-migration-for-bridged-pod-network.md:25]: What's the criteria for this feature to graduate from a feature flag to being default enabled?

Comment (davidvossel) [reply design-proposals/live-migration-for-bridged-pod-network.md:25]: >What's the criteria for this feature to graduate from a feature flag to being default enabled?

it would need to function for the majority of guest OSs and not cause undesirable behavior for existing installs. I'm unsure if this will reach a point where we can make it on by default, but it's a possibility.

**Comment by davidvossel** [comment]
>  I answered the most of conversation threads, not sure, what I suppose to do now?

@AlonaKaplan @maiqueb @EdDev Given the design outlined here, do you all have any reservations about introducing this feature behind a feature gate now? I understand there's a desire to create a new abstraction for network bindings, but unless there's a predictable time frame for when this work will be completed, I'd hate to make Andrei wait around. His logic doesn't appear to be too terribly invasive to the existing binding.

**Comment by maiqueb** [comment]
FWIW - I am not a maintainer - I do not have any reservation about the proposed code.

What I am concerned with is adding more stuff to bridge binding - a binding that afaiu - relies on an hack to work when IPAM is involved - I am referring to the dummy interface switcharoo we perform with that binding.

Reading the [CNI spec](https://github.com/containernetworking/cni/blob/main/SPEC.md#check-check-containers-networking-is-as-expected) for the **CHECK** command:
> The plugin should return an error if a resource included in the CNI Result type (interface, address or route) was created by the plugin, and is listed in prevResult, but is missing or in an invalid state.

AFAIU, when we move the IP from the pod interface to the dummy we break the above rule - the CNI result says the IP address is in iface index X when  in truth said IP is in iface index Y.

Whenever the CRI fixes this (I consider this a bug), I assume bridge binding with IPAM will break. Unsure on how to fix it. 

I'm aware this is **not** something introduced in this PR, and as such, holding the PR hostage over it shouldn't be done.

Now, as @davidvossel said, we want to redesign the network binding abstraction to allow users to provide their own bindings outside of KubeVirt. I still feel this is **the** right way to achieve this feature. Then again, I cannot provide any time frame for this atm. We haven't even begun planning it. 

As such, my opinion is we shouldn't hold this community effort. Having said that, I am not a maintainer, and this is not my call to make.

**Comment by AlonaKaplan** [comment]
> What I am concerned with is adding more stuff to bridge binding - a binding that afaiu - relies on an hack to work when IPAM is involved - I am referring to the dummy interface switcharoo we perform with that binding.
> 
> Reading the [CNI spec](https://github.com/containernetworking/cni/blob/main/SPEC.md#check-check-containers-networking-is-as-expected) for the **CHECK** command:
> 
> > The plugin should return an error if a resource included in the CNI Result type (interface, address or route) was created by the plugin, and is listed in prevResult, but is missing or in an invalid state.
> 
> AFAIU, when we move the IP from the pod interface to the dummy we break the above rule - the CNI result says the IP address is in iface index X when in truth said IP is in iface index Y.
> 
> Whenever the CRI fixes this (I consider this a bug), I assume bridge binding with IPAM will break. Unsure on how to fix it.
> 

I totally agree with Miguel. I consider bridge binding as hacky and I would prefer not to add more logic to it.
As I mentioned before, we didn't add anything to this binding for years. We didn't even add IPv6 support to it. And I believe the general direction is to get rid (keep it for backward compatibility only) of this binding (for primary network) and substitute it with another binding, maybe [passt](https://kubevirt.io/user-guide/virtual_machines/interfaces_and_networks/#passt).

With passt the VM can have the same IP as the pod, but the performance won't be as good as with bridge binding.
However, if the issue is the performance maybe masquerade with multi-queue can be enough.

Another option until the redesign is done is a hook point as I suggested [here](https://github.com/kubevirt/community/pull/182#issuecomment-1196768703).
@kvaps did you have a change to try it?

**Comment by rthallisey** [comment]
/remove-lifecycle stale

@kvaps I think this is close.  Do you plan on doing another revision?

**Comment by kvaps** [comment]
This patch works nice in our fork.

It can't be implemented using hooks, as KubeVirt blocking all the migrations with bridge binding on pod networking.

So I'm waiting for plug-in interface to implement this as external plugin for KubeVirt.

The design of this feature is not going to change any way, so I'd like to keep this proposal open

**Comment by xagent003** [comment]
I would like this too... we are using Calico with a dedicate IPPool for VMs and specifying a fixed IP from this via a calico annotation. So VM has same IP as virt-lauincher Pod, and this IP is always the same even if it is recreated on a new node.

We don't want masquerade because of extra added NAT even between 2 VMs talking on same network, and and VMs won't know their actual IP configured inside.

With Multus, and a L2 model and DHCP and kubemacpool, this is fine for some cases but doesn't let us scale network creations as it requires us to tag infra and add physical gateways for each L2 network. We tried kube-ovn to allow virtual networks, but hit many issues, apparently it doesn't work well and isnt tested as a secondary CNI nor with Kubevirt. Multus networks also aren't first class K8s citizens and there is no firewalling/security group features really

So, we'd really like to use a primary CNI like Calico for its feature set and L3 networking, use a fixed IP for VMs, and have it be in bridge mode so VM doesn't have private IP and extra NAT. Why would live migration be an issue in this case?

**Comment by xuzhenglun** [comment]
Is anybody know why a pod network with bridge mode would be a problem? or why multus with bridge works fine. what if I use multus default interface with bridge mode?

```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  labels:
    kubevirt.io/os: linux
  name: test-vm
  namespace: kubevirt
spec:
  runStrategy: Manual
  template:
    spec:
      domain:
        cpu:
          cores: 2
        devices:
          disks:
          - disk: {}
            name: disk0
          interfaces:
          - bridge: {}
            name: default
        machine:
          type: q35
        resources:
          requests:
            memory: 8192M
      networks:
      - multus:
          default: true
          networkName: calico-default
        name: default
...

```

**Comment by kvaps** [comment]
I migrated my fork from `kvaps/community` to `deckhouse/3p-kubevirt-community`, now it will be supported by deckhouse project
fyi

**Comment by aburdenthehand** [comment]
@kvaps Heya - now that plugins have landed, what's the state of this proposal, and do you need something to get it over the line?

**Pull request closed without merging**
